### DISCUSSION TITLE
Year 2022, Season 4

### DISCUSSION THREAD

**Discussion by AllanChain** [AllanChain/micro-blog#5] [MicroBlog]
Time Fies!

**Comment by AllanChain** [comment]
之前只是将代际更替用来理解生物进化和文化变迁之类，现在发现用在装 Linux 和朝代的更替也是可行。

有的软件可以不用重装系统，但要更改文件系统之类，就算不重装，折腾程度也差不多。—— 王朝的根基即使有问题也不可动摇；“改良”还是“革命”

但不到硬件老化坏掉或操作不当系统滚挂得时候，是万没有动力去重装系统的。—— Fall and rebuild

要让下一个系统更好，现在就需要不断探索研究、扩展眼界。否则在装新系统的短时间内再想是不会有一个很好的方案的。—— 研究政治（儒学）在什么时候都是重要的

**Comment by AllanChain** [comment]
WTFs on Julia 👇

**Reply by AllanChain** [reply]
Julia might be fast, but its JIT compilation is extremely slow. For example, JuliaFormatter takes 20 sec to format a file, with most of the time spent on JIT compilation. (https://github.com/domluna/JuliaFormatter.jl/issues/633)

**Reply by AllanChain** [reply]
Julia's `mean(a, dims = 2)` is equivalent to `np.mean(a, axis=1, keepdims=True)`, and there is no way to set `keepdims=False` and you must do it separately. See https://github.com/JuliaLang/julia/issues/16606.

**Reply by AllanChain** [reply]
Not to mention that Julia has a weird syntax for creating arrays: [docs link](https://docs.julialang.org/en/v1/manual/arrays/#man-array-concatenation)

**Reply by AllanChain** [reply]
To add a new element to an array/list/vector:
- JavaScript: `a.push(e)`
- Python: `a.append(e)`
- Julia: `push!(a, e)`

To add elements in one array/list/vector to another:
- Javascript: `a.push(...b)`
- Python: `a.extend(b)`
- Julia: `append!(a, b)`

**Reply by AllanChain** [reply]
Julia chooses to use `include` instead of modules. File structures are programmatically meaningless, and you are expected to assemble all the files in `MyProject.jl`.

**Reply by AllanChain** [reply]
Oh, I forgot to mention the obscure error messages.

**Reply by AllanChain** [reply]
Julia doesn't support ellipsis for array slicing, and the third-party implementation also have a lot limitations, which makes it very inconvenient to slice arrays with  unknown dimension.

**Reply by AllanChain** [reply]
`ceil(256/6)` returns a float. You need to use `ceil(Int64, 256/6)` instead for integer.

**Comment by AllanChain** [comment]
当我们开始讨论大局的时候，就意味着有人要倒霉了。至于大局最终能否保全却是次要的，先开始倒霉再说。

**Comment by AllanChain** [comment]
可以这样评价软件: feature 不足，bug 有余

**Reply by Yixuan-Wang** [reply]
```python
for bug in bugs: assert bug in features
```

**Comment by AllanChain** [comment]
各自在各自的位置，各自算自己的那本账，果真能团结一致行动，也算是奇迹了。

可不管如何努力，我们似乎永远也摆脱不了前半句的现实。

**Comment by AllanChain** [comment]
说不定英文的内容可以往 Fediverse 上放？那这里可能就只剩下一些 nonsense 了。

**Comment by AllanChain** [comment]
- 程序员：这个调音软件一定能帮助很多音乐家吧！
- 我：给五根橡皮筋调音
- 还是我：看看自己唱歌有多跑调

**Reply by AllanChain** [reply]
还是我：才发现人声和口哨的音域没有半点交集

**Comment by AllanChain** [comment]
胡乱发言: 教务老师之所以能一眼看出来一张照片是不是大学时期拍的，不是看你长大了多少，而是看你眼里有没有被大学生活摧残的痕迹。

**Comment by AllanChain** [comment]
当你发现一个不可思议的行为是由于同一个库不同函数的默认单位不同而导致的时候：

```python
>>> from pyscf import gto
>>> gto.M(atom="H 0 0 0; H 1.4 0 0", basis="sto-3g").eval_gto("GTOval_sph", [[0.7, 0, 0]])
array([[0.32583116, 0.06929586]])
>>> gto.M(atom="H 0 0 0; H 1.4 0 0", basis="sto-3g", unit="B").eval_gto("GTOval_sph", [[0.7, 0, 0]])
array([[0.32583116, 0.32583116]])
```

<p align=center><img src="https://user-images.githubusercontent.com/36528777/201502511-b3d216a5-6631-4c85-aeb1-16a0df6c415c.png" width="200"></p>

**Comment by AllanChain** [comment]
门外有门，筛外有筛——指关严实了一遍遍过筛

**Comment by AllanChain** [comment]
都是用阴阳历、过阴阳节的人，就不要争谁阴阳谁不阴阳了

**Comment by AllanChain** [comment]
球进了，但当事守门员是最后一个意识到的。(与世界杯无关)

**Comment by AllanChain** [comment]
macOSoft
